### PR TITLE
Defcn

### PR DESCRIPTION
**Pull request by falex-ml**
major refactoring reflecting the fact that we're not FCN-only anymore...

### CODE REVIEW

File: arch.py [hunk 1]
```diff
@@ -1,7 +1,9 @@
+import sys
 import tensorflow as tf
 from utils.upsampling import bilinear_upsample_weights
 
-# ..assumes slim_models is added to path
+# assuming our fork of tf-models is cloned side-by-side with current repo
```
Comment (grobman) [inline arch.py]: language fix: assuming tf_models_hailofork. not our fork..

File: test.py [hunk 2]
```diff
@@ -231,24 +235,24 @@ def main(args):
...
+        netclass = arch.FcnArch
     print("Using architecture " + netclass.__name__)
 
-    args = fcn_train.resolve_dataset_family(args)
+    args = train.resolve_dataset_family(args)
     args.num_images = args.num_images or args.total_val_images
-    fcn_builder = netclass(number_of_classes=args.num_classes, is_training=False, net=args.basenet,
+    net_builder = netclass(number_of_classes=args.num_classes, is_training=False, net=args.basenet,
                            trainable_upsampling=args.trainable_upsampling, fcn16=args.fcn16, fcn8=args.fcn8)
     # fcn_builder = fcn_arch.FcnArch(number_of_classes=number_of_classes, is_training=False, net=args.basenet,
```
Comment (grobman) [inline test.py]: remove commentned code

File: test.py [hunk 3]
```diff
@@ -231,24 +235,24 @@ def main(args):
...
     #                                trainable_upsampling=args.trainable_upsampling, fcn16=args.fcn16)
 
     # ..From logits to class predictions
     if pixels > 0 and pixels == (pixels/32)*32.0 :
         def fcnfunc_img2labels(img):
```
Comment (grobman) [inline test.py]: shouldn't the function name be generic, e.g. netfunc_img2labels?

File: utils/tf_records.py [hunk 4]
```diff
@@ -4,10 +4,14 @@
 # to actual rgb values. This is specific to PASCAL VOC
```
Comment (grobman) [inline utils/tf_records.py]: I think the name tf_records.py is somewhat non descriptive

File: utils/tf_records.py [hunk 5]
```diff
@@ -225,31 +230,75 @@ def tfrecordify_coco_stuff_things(imgsdir = '/data/coco',
     write_image_annotation_pairs_to_tfrecord(filename_pairs=valpairs,
                                              tfrecords_filename=imgsdir+'/validation.tfrecords')
 
-def tfrecordify_camvid(rootdir = '/data/camvid'):
+def tfrecordify_camvid(datadir = '/data/camvid'):
     '''
         Assumes stuff is e.g. cloned from https://github.com/alexgkendall/SegNet-Tutorial // CamVid
```
Comment (grobman) [inline utils/tf_records.py]: Assumes stuff is e.g?

File: utils/tf_records.py [hunk 6]
```diff
@@ -225,31 +230,75 @@ def tfrecordify_coco_stuff_things(imgsdir = '/data/coco',
...
          - 11-class version encoded as the usual #class (not color-code)
 
         TODO 32-class version
     '''
     import os
```
Comment (grobman) [inline utils/tf_records.py]: imports shouldn't be in function

File: utils/tf_records.py [hunk 7]
```diff
@@ -4,10 +4,14 @@
 # to actual rgb values. This is specific to PASCAL VOC
 # dataset data. If you don't want thit type of behaviour
```
Comment (grobman) [inline utils/tf_records.py]: line 201 - please remove "WTF"... 
Also...please search other instances and remove. We're a pg-13 company :)

File: utils/tf_records.py [hunk 8]
```diff
@@ -225,31 +230,75 @@ def tfrecordify_coco_stuff_things(imgsdir = '/data/coco',
...
+    datapath = args.datapath if args.datapath != '' else '/data/'+args.dataset_family
+    
+    if args.dataset_family=='camvid':         
+        tfrecordify_camvid(datapath)
+    elif args.dataset_family=='coco':
+        tfrecordify_coco_stuff_things(datapath, 'stuffthings_pixellabels')
+    elif args.dataset_family=='pascal_seg':
+        voc_path = args.voc_path if args.voc_path != '' else os.path.join(datapath, 'VOCdevkit/VOC2012')
+        sbd_path = os.path.join(datapath, 'benchmark_RELEASE')
+        tfrecordify_pascal_seg(voc_path, sbd_path, datapath)
```
Comment (grobman) [inline utils/tf_records.py]: why is datapath in VOC treated differently (with explicit code) than for COCO/Camvids?

File: test.py (hunk 2 above)
Comment (grobman) [reply test.py]: @falex-ml

File: utils/tf_records.py (hunk 8 above)
Comment (falex-ml) [reply utils/tf_records.py]: since in PASCAL images and labels generally travel alone, 
there's a basic image devkit that you may possibly have already from obj-det projects (As we do).
